### PR TITLE
Configured maven bundle plgin to produce osgi bundles

### PR DESCRIPTION
**Pull request by ieugen**
* Fixes https://github.com/atlassian/commonmark-java/issues/128 
* Over the next days I will test the bundles and see if any changes from the default are needed
* Waiting for a release :)

### CODE REVIEW

**Comment by ieugen** [comment]
After this I will start working on a simple Karaf feature to easily install the bundles in Karaf.

File: commonmark/pom.xml [hunk 1]
```diff
@@ -8,6 +8,7 @@
     </parent>
 
     <artifactId>commonmark</artifactId>
+    <packaging>bundle</packaging>
```
Comment (robinst) [inline commonmark/pom.xml]: I'm a bit uncomfortable with changing the packaging type here. Is there a way to add the manifest entries without changing this?

File: commonmark/pom.xml [hunk 2]
```diff
@@ -42,6 +43,19 @@
...
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                  <instructions>
+                    <Export-Package>
+                      *,
+                      org.commonmark.internal
+                    </Export-Package>
```
Comment (robinst) [inline commonmark/pom.xml]: What about the other artifacts, don't they need something like this as well? They also have `internal` packages that should not be exported.

File: commonmark/pom.xml [hunk 3]
```diff
@@ -42,6 +43,19 @@
                     </archive>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
```
Comment (robinst) [inline commonmark/pom.xml]: What does this do, is it needed?

File: commonmark/pom.xml (hunk 1 above)
Comment (ieugen) [reply commonmark/pom.xml]: Actually there is, but adds more configuration. However there are no issues with bundle packaging. The plugin exists for a lot of time and works well with all other plugins.

[1] https://bnd.bndtools.org/tools/felix-maven.html

File: commonmark/pom.xml (hunk 2 above)
Comment (ieugen) [reply commonmark/pom.xml]: I haven't got to using extensions yet. If those internal packages are not used by other bundles then no need to export them. I think the whiteboard pattern might fit for extensions. Using that pattern extensions can make themselves available to any user. 

[1] https://en.wikipedia.org/wiki/Whiteboard_Pattern

File: commonmark/pom.xml (hunk 3 above)
Comment (ieugen) [reply commonmark/pom.xml]: This plugin updates the MANIFEST.MF file with osgi metadata.

File: commonmark/pom.xml (hunk 2 above)
Comment (robinst) [reply commonmark/pom.xml]: They are not used by other bundles, yeah.

File: commonmark/pom.xml (hunk 3 above)
Comment (robinst) [reply commonmark/pom.xml]: I meant what does the `<extensions>true</extensions>` do?

Comment (ieugen) [reply commonmark/pom.xml]: Yes it is needed. Please see docs: https://maven.apache.org/guides/mini/guide-using-extensions.html .

That is how the plugin configuration is done for it to work.

**Comment by ieugen** [comment]
> Does it automatically use "-" for the non-groupId part?
The algorithm for choosing a name is described in the docs. It works very well. Symbolic name is used to identify the bundle by the framework and for humans. If you are not happy with it you can change it. I don't ever recall changing that in all my work. 

Imports and exports usually require some attention. Other than that, defaults work very well 99% of the time for me.

**Comment by robinst** [comment]
Thanks for your patience with this. I'll merge this if:

* You make it work without changing `<packaging>` (I don't want to make it harder to set up for any IDEs, etc)
* You also go through the extensions to add the appropriate `Export-Package` there

> The algorithm for choosing a name is described in the docs. It works very well. Symbolic name is used to identify the bundle by the framework and for humans. If you are not happy with it you can change it. I don't ever recall changing that in all my work.

Ok, let's leave it. Just looked a bit weird because it only got a `-` at the end :).

**Comment by josephw** [comment]
>     * You make it work without changing `<packaging>`

FWIW, the example in  the [Apache Felix Maven Bundle Plugin documentation](https://bnd.bndtools.org/tools/felix-maven.html#real-world-example) shows that this is a relatively small change:

    <plugin>
      <artifactId>maven-jar-plugin</artifactId>
      <configuration>
        <archive>  
          <manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>
        </archive> 
      </configuration>
    </plugin>  
    <plugin>   
      <groupId>org.apache.felix</groupId>
      <artifactId>maven-bundle-plugin</artifactId>
      <executions>
        <execution>
          <id>bundle-manifest</id>
          <phase>process-classes</phase>
          <goals>    
            <goal>manifest</goal>
          </goals>   
        </execution>
      </executions>
    </plugin>

With this project's module structure, adding that once to `commonmark-parent` may be enough to get OSGi bundles without modifying the packaging type.

**Comment by robinst** [comment]
Closing this. Open to a less invasive PR that does it the way @josephw pointed out.

**Pull request closed without merging**